File: leetcode/213-house-robber-ii/main.go
package main

/*
	1st approach: reuse leetcode 213
	- since adjacent means only the i-1 and i+1, problem becomes to rob either
		ouse[0]-House[n-2] or House[1]-House[n-1], depending on which choice offers more money

	Time		O(2n)
	Space		O(1)
	0 ms, faster than 100.00%
*/
func rob(nums []int) int {
	if len(nums) == 0 {
		return 0
	}
	if len(nums) == 1 {
		return nums[0]
	}
	if len(nums) == 2 {
		return max(nums[0], nums[1])
	}
	a := notCicular(nums[:len(nums)-1])
	b := notCicular(nums[1:])
	return max(a, b)
}

func notCicular(nums []int) int {
	prevprev := 0
	prev := 0
	for i := 0; i < len(nums); i++ {
		cur := nums[i]
		temp := max(prevprev+cur, prev)
		prevprev = prev
		prev = temp
	}
	return prev
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {

}

File: leetcode/213-house-robber-ii/main.py
"""
    1st approach: reuse leetcode 198
	- circular means that we cannot rob houses[0] and houses[n-1] at the same time, so the result is either:
        - from houses[1] to houses[n-2]
        - Or from houses[2] to houses[n-1]

	Time		O(2n)
	Space		O(1)
	8 ms, faster than 99.45%
"""


class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums) == 0:
            return 0
        if len(nums) == 1:
            return nums[0]
        a = self.simpleRob(nums[:-1])
        b = self.simpleRob(nums[1:])
        return max(a, b)

    def simpleRob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        rob = 0
        notRob = 0
        for x in nums:
            temp = rob
            rob = max(rob, notRob + x)
            notRob = temp
        return max(rob, notRob)
